Add segmented aggregation in AggregationNode

kewang1024 opened the pull request:
```
== NO RELEASE NOTE ==
```

If the grouped-by keys contains elements from the prefix of the sorted-by key, we can enable segmented aggregation 

For example:the table is sorted by F1, F3 and and we do Group by F1, F2
F2 is not sorted, so we can’t do streaming aggregation for each <F1, F2> group; however since F1 is sorted, we segment the data by F1’s value, for example the first segment, F1’s values are all a, now we can build a hashtable for each segment and do calculation and flush the data once a segment is finished

so it still saves CPU because we don’t do look up for F1, and the result hash table we keep is also smaller compared to the full hash table
<img width="565" alt="image" src="https://user-images.githubusercontent.com/50636602/165151336-a5249301-4362-480d-9294-a96e505d84cf.png">

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -325,6 +325,16 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
                 preGroupedSymbols = groupingKeys;
             }
+            else {
+                // Enable segmented streaming aggregation if the input is pre-grouped by GroupBy key prefix.
```
This requirement might be too strong. It is not necessary for the input to be grouped on the prefix of the grouping keys. It is sufficient for the input to be grouped on a subset of the grouping keys.

yuanzhanhku replied:
That is right. As long as the prefix of the sorted columns is a subset of the group by column, we can enable segmented aggregation.

The problem here is that presto table writer doesn't support group by property, so we have to create the table with sort_by columns and infer the groupby property. In this case, the rows are grouped by the prefix of the sort_by columns. We don't want to expose all prefix column combinations in the data property. Only the full sort_by column list is in the sort property.

The logic below need to be changed to iterate all possible prefixes of the sort_by columns in child.getProperties().getLocalProperties() and see if it is a subset of the groupingKeys. And we need to find the longest prefix of the sort_by columns that is a subset of the groupingKeys.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:
```diff
@@ -594,6 +594,9 @@ public Void visitAggregation(AggregationNode node, Void context)
                 type = format("(%s)", node.getStep().toString());
             }
+            if (node.isSegmentedStreamable()) {
+                type = format("%s(SEGMENTED STREAMING)", type);
```
Perhaps, we could include the preGroupedKeys in the output in case node.isStreamable() is false.

yuanzhanhku commented on presto-spi/src/main/java/com/facebook/presto/spi/plan/AggregationNode.java:
```diff
@@ -210,7 +210,12 @@ public PlanNode replaceChildren(List<PlanNode> newChildren)
 
     public boolean isStreamable()
```
We need to check if any of the callsite also applies to the SegementedAggregation case.
For example, PushPartialAggregationThroughJoin::isSupportedAggregationNode returned false for streaming aggregation. There is also a usage in HashGenerationOptimizer.java. I haven't check the details about these rules yet. @mbasmanova, do you know what should be the expected value for segmented aggregation?

kewang1024 replied:
1. isStreamable is used in two places out of all the optimizer rules, similarly we add isSegmentedAggregationEligible in those places as well
- HashGenerationOptimizer
For segmented aggregation, we shouldn't calculate hash for all the fields, but only for the fields that need to compute hash, this requires more changes and benchmark result to support such optimization, left a todo for future support 

- PushPartialAggregationThroughJoin
It only pushes Partial aggregation, for Partial aggregation, we currently haven't added segmented aggregation, so we should be safe there

2. Regarding the aggregation pushdown, there are two places
-  PushPartialAggregationThroughExchange
According to the comments, we would anyway drop the property, so for segmented aggregation, the original behavior is not affected
// preGroupedSymbols reflect properties of the input. Splitting the aggregation and pushing partial aggregation
// through the exchange may or may not preserve these properties. Hence, it is safest to drop preGroupedSymbols here.

-  PushPartialAggregationThroughJoin
Explained above

yuanzhanhku commented on presto-hive/src/test/java/com/facebook/presto/hive/TestSegmentedAggregation.java:
```diff
@@ -0,0 +1,91 @@
+                    "SELECT custkey, name, nationkey, COUNT(*) FROM test_segmented_streaming_customer \n" +
+                            "WHERE ds = '2021-07-11' GROUP BY 1, 2, 3",
+                    anyTree(aggregation(
+                            singleGroupingSet("custkey", "name", "nationkey"),
```
Please add the following tests 
1. with a different order of group by columns:
     sorted_by = ARRAY['custkey', 'name']; singleGroupingSet("name", "custkey", "nationkey")
     expected pregrouped fields: custkey. But this may require additional implementation to support it. We can add a TODO to support this later. See also the comment below. 
2. sorted_by has extra columns which are not in the group by columns. 
    sorted_by = ARRAY['custkey', 'name']; singleGroupingSet("custkey", "nationkey")
    expected pregrouped fields: custkey

yuanzhanhku commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -335,6 +335,19 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
+                // Enable segmented aggregation if the prefix of the sorted-by columns is a subset of the group by column.
+                // Use the longest prefix of the sorted-by columns that satisfy the requirement to maximize the segment aggregation's preGroupedSymbols
+                for (int i = child.getProperties().getLocalProperties().size(); i > 0; i--) {
+                    List<LocalProperty<VariableReferenceExpression>> sortedByKeysPrefix = child.getProperties().getLocalProperties().subList(0, i);
```
getLocalProperties() may contain non-sortedBy property. We need to filter it.
Besides, groupby property could also work. It doesn't have to be sorted by property. We can support only sortedBy property for now and add a TODO to support groupby property later.

yuanzhanhku commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -335,6 +335,19 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
+                // Use the longest prefix of the sorted-by columns that satisfy the requirement to maximize the segment aggregation's preGroupedSymbols
+                for (int i = child.getProperties().getLocalProperties().size(); i > 0; i--) {
+                    List<LocalProperty<VariableReferenceExpression>> sortedByKeysPrefix = child.getProperties().getLocalProperties().subList(0, i);
+                    List<Optional<LocalProperty<VariableReferenceExpression>>> matchResult = LocalProperties.match(sortedByKeysPrefix, LocalProperties.grouped(groupingKeys));
```
This may not work if groupingKeys has a different order than the sortedByKeys. We may need to rearrange the order of the fields in groupingKeys according to the order in sortedByKeysPrefix. We can add a TODO to support this case later.

yuanzhanhku commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -331,10 +331,20 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
+            List<LocalProperty<VariableReferenceExpression>> sortedByKeys = child.getProperties().getLocalProperties();
+            if (!LocalProperties.match(sortedByKeys, LocalProperties.grouped(groupingKeys)).get(0).isPresent()) {
                 // !isPresent() indicates the property was satisfied completely
                 preGroupedSymbols = groupingKeys;
```
I think we also need to check the streaming partition columns and make sure it is partitioned by the group columns across splits. IIRC, I didn't add this check in my prototype because it was implicitly checked. If the streaming aggregation doesn't match the group columns, a local exchange will be added below the aggregation and the local property will not match. We'd better document this assumption in the code comment.

However, the merge join planning has to check the partitioning columns across splits and ensure colocated join is planned. Only checking the local property will not work for merge join. (This is not related to this PR. But need to fix in the merge join planning code).

yuanzhanhku commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -331,10 +331,20 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
+            else {
+                // Enable segmented aggregation if the prefix of the sorted-by keys is a subset of the grouped-by keys
+                // Use the longest prefix to maximize the segment aggregation's preGroupedSymbols
+                List<Optional<LocalProperty<VariableReferenceExpression>>> matchResult = LocalProperties.match(sortedByKeys, LocalProperties.grouped(groupingKeys));
```
This match function is really confusing. We should improve its function comments. :)
What does it return if columns in groupingKeys have a different order from sortedByKeys?

yuanzhanhku commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -331,10 +331,25 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
+            // [A, B] [(B, A)]   ->   List.of(Optional.empty())
+            // [A, B] [B]        ->   List.of(Optional.of(GroupingProperty(B)))
+            // [A, B] [A]        ->   List.of(Optional.empty())
+            // [A, B] [A, C]     ->   List.of(Optional.of(GroupingProperty(C)))
```
The match function seems very powerful. :) Very clean solution. Thanks for figuring it out.

Could you please add another example for
[A, B] [D, A, C] (basically the first group by column is not part of the sort by columns, this case is not included in any of the above examples)
In this case will it return [D, C] and we can set A in the pre-grouped columns?
Please add a test for this case as well.

yuanzhanhku reviewed (approved):
LGTM. Thanks for implementing this feature!

kewang1024 commented:
@rschlussel gentle ping just incase you miss the notification

rschlussel commented on presto-spi/src/main/java/com/facebook/presto/spi/plan/AggregationNode.java:
```diff
@@ -210,7 +210,12 @@ public PlanNode replaceChildren(List<PlanNode> newChildren)
     public boolean isStreamable()
     {
-        return !preGroupedVariables.isEmpty() && groupingSets.getGroupingSetCount() == 1 && groupingSets.getGlobalGroupingSets().isEmpty();
+        return !preGroupedVariables.isEmpty() && groupingSets.getGroupingSetCount() == 1 && groupingSets.getGlobalGroupingSets().isEmpty() && preGroupedVariables.size() == groupingSets.groupingKeys.size();
```
This line is kind of long. I would split it up across multiple lines for readability

rschlussel commented on presto-spi/src/main/java/com/facebook/presto/spi/plan/AggregationNode.java:
```diff
@@ -210,7 +210,12 @@ public PlanNode replaceChildren(List<PlanNode> newChildren)
+
+    public boolean isSegmentedAggregationEligible()
+    {
+        return !preGroupedVariables.isEmpty() && groupingSets.getGroupingSetCount() == 1 && groupingSets.getGlobalGroupingSets().isEmpty() && preGroupedVariables.size() < groupingSets.groupingKeys.size();
```
ditto

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -350,6 +366,17 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
+        private List<VariableReferenceExpression> extractPreGroupedSymbols(List<VariableReferenceExpression> groupingKeys, Set<VariableReferenceExpression> exludedColumns)
+        {
+            ImmutableList.Builder<VariableReferenceExpression> preGroupedSymbolsBuilder = ImmutableList.builder();
+            for (VariableReferenceExpression groupingKey : groupingKeys) {
```
we usually prefer stream apis when we can because it's more succict:
```
return groupingKeys.stream()
            .filter(groupingKey -> !excludedColumns.contains(groupingKey))
            .collect(toImmutableList())
```

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -350,6 +366,17 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
             return deriveProperties(result, child.getProperties());
         }
 
+        private List<VariableReferenceExpression> extractPreGroupedSymbols(List<VariableReferenceExpression> groupingKeys, Set<VariableReferenceExpression> exludedColumns)
```
typo: exludedColumns -> excludedColumns (missing 'c' in "excluded")

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -331,10 +331,26 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
 
             List<VariableReferenceExpression> preGroupedSymbols = ImmutableList.of();
-            if (!LocalProperties.match(child.getProperties().getLocalProperties(), LocalProperties.grouped(groupingKeys)).get(0).isPresent()) {
+            // Logic in LocalProperties.match(localProperties, groupingKeys)
```
thanks for the comment here- was very helpful!

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java:
```diff
@@ -166,7 +166,8 @@ public PlanWithProperties visitAggregation(AggregationNode node, HashComputation
             List<VariableReferenceExpression> groupingKeys = node.getGroupingKeys();
-            if (!node.isStreamable() && !canSkipHashGeneration(node.getGroupingKeys())) {
+            if (!node.isStreamable() && !node.isSegmentedAggregationEligible() && !canSkipHashGeneration(node.getGroupingKeys())) {
+                // todo: for segmented aggregation, add optimizations for the fields that need to compute hash
```
seems simple to filter out the pregrouped columns, so why not do this already as part of this PR?

rschlussel reviewed (commented):
I haven't reviewed the tests yet because I was a bit confused about them. Is this feature available already on the execution side?

kewang1024 commented:
> [WIP] Add segmented streaming aggregation support during execution #17618

No, this is the worker side change: https://github.com/prestodb/presto/pull/17618

kewang1024 commented:
> I haven't reviewed the tests yet because I was a bit confused about them. Is this feature available already on the execution side?

No, this is the worker side change: https://github.com/prestodb/presto/pull/17618

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java:
```diff
@@ -166,7 +166,8 @@ public PlanWithProperties visitAggregation(AggregationNode node, HashComputation
             Optional<HashComputation> groupByHash = Optional.empty();
             List<VariableReferenceExpression> groupingKeys = node.getGroupingKeys();
-            if (!node.isStreamable() && !canSkipHashGeneration(node.getGroupingKeys())) {
+            if (!node.isStreamable() && !node.isSegmentedAggregationEligible() && !canSkipHashGeneration(node.getGroupingKeys())) {
```
we should only skip hash generation if segmented aggregation is enabled otherwise there will be a regression until the execution side is ready

rschlussel reviewed (commented):
I think you need to add a session property specific to segmented aggregation that is disabled by default for now.  Otherwise we will have a regression while we wait for the execution side to be ready.

kewang1024 commented:
Thanks for the review @rschlussel 
I addressed most of the comments, for the rest
1. We left a todo in `HashGenerationOptimizer` because the segmented aggregation worker implementation hasn't been completed, it depends on the implementation how we want to pre-compute the hash for segmented aggregation plan
2. I added one additional session property to guard this feature from being generated in the beginning, so the rest of places won't need to be guarded anymore

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -1180,6 +1181,11 @@ public SystemSessionProperties(
+                booleanProperty(
+                        SEGMENTED_AGGREGATION_ENABLED,
+                        "Enable segmented aggregation.",
+                        false,
```
Since after the worker side implementation is complete, we can remove this session property, that's why I didn't create the config accordingly

rschlussel replied:
I think it's probably good to have a config because we might continue to use it even after the worker side implementation is complete e.g. if there's a bug somewhere, it would be good to have a shutoff switch.

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -1180,6 +1181,11 @@ public SystemSessionProperties(
+                        SEGMENTED_AGGREGATION_ENABLED,
+                        "Enable segmented aggregation.",
+                        false,
+                        true),
```
yes, setting it to true means hidden

rschlussel replied:
ah you're right. got them mixed up!
